File: tslib-py/tslib/stationarity.py
from typing import List, Optional

from ._http import _Session
from .models import AdfResult, KpssResult


class StationarityAPI:
    def __init__(self, session: _Session) -> None:
        self._s = session

    def adf(self, data: List[float], *, lag: Optional[int] = None) -> AdfResult:
        body: dict = {"data": data}
        if lag is not None:
            body["lag"] = lag
        return AdfResult._from(self._s.post("/api/tests/adf", body))

    def kpss(
        self,
        data: List[float],
        *,
        regression_type: str = "LEVEL",
        lags: Optional[int] = None,
    ) -> KpssResult:
        body: dict = {"data": data, "regressionType": regression_type}
        if lags is not None:
            body["lags"] = lags
        return KpssResult._from(self._s.post("/api/tests/kpss", body))

File: tslibR/R/stationarity.R
#' Augmented Dickey-Fuller (ADF) test
#'
#' Tests the null hypothesis that a unit root is present (series is
#' non-stationary).
#'
#' @param client A [`tslib_client`] object.
#' @param data Numeric vector of observations.
#' @param lag Optional number of lags. If `NULL` the server selects the lag
#'   automatically.
#' @return Named list with `statistic`, `pValue`, `lag`, `stationary` (bool),
#'   and `needsDiff` (bool).
#' @export
tslib_adf <- function(client, data, lag = NULL) {
  body <- .compact(list(data = as.list(data), lag = lag))
  .tslib_post(client, "/api/tests/adf", body)
}

#' KPSS stationarity test
#'
#' Tests the null hypothesis that a series is stationary (opposite of ADF).
#' Use together with ADF for a confirmatory pair of tests.
#'
#' @param client A [`tslib_client`] object.
#' @param data Numeric vector of observations.
#' @param regression_type Regression type: `"LEVEL"` (default) tests
#'   level-stationarity; `"TREND"` tests trend-stationarity.
#' @param lags Optional number of lags for the long-run variance estimator.
#' @return Named list with `statistic`, `lags`, `regressionType`, critical
#'   values, and stationarity flags at 1 % and 5 % levels.
#' @export
tslib_kpss <- function(client, data,
                       regression_type = "LEVEL",
                       lags            = NULL) {
  body <- .compact(list(
    data            = as.list(data),
    regressionType  = regression_type,
    lags            = lags
  ))
  .tslib_post(client, "/api/tests/kpss", body)
}
